### DISCUSSION TITLE
Next Moves

### DISCUSSION THREAD

**Discussion by xpollen8** [xpollen8/jazzbutcher.com#141] [Ideas]
*Taps mic*

Ok, everybody.

V2 has been up for awhile, but it is more a V1.5, as the stated longterm goals have yet to be reached: portable data export with schema for others to run with.

V2 instead spent a *TON* of time in data clean-up, clarification, and the incorporation of more press articles, side-projects and pre-JBC material.

While the majority of the "data" (albums, gigs, lyrics, people, press, etc) still uses a SQL database as the master source, and is accessible via an API..  The codebase now also publishes JSON representations of that data in a static manner.  In that sense, if the database ever goes down due to the demise of the bill-payer.. at least the data is now public as JSON.  So, we've gained some safety.

But we're not yet at the point where someone can download a tarball and - ta-da - have everything JBC on their local machine.

----------

Here's what I've been mulling over.

JBC com is *mostly* a static website.  The only thing right now that needs to be dynamic at this point is the comments feature (and search).

Everything else could be a static website.

How that site gets statically generated is the question.

One semi-sane way to do this is to store the data bits that make up a page as markdown files, and then a pipeline (jekyl) notices changes which trigger the build of the resulting page.  Github Pages would be a natural solution here, as all this comes for free.

I like this idea because the easily human-editable markdown files become *The True Source* for each "thing".
Markdown files could easily be translated into anything anyone else might want: turn 'em into XML, JSON, SQL, etc.

After having driven down *The React Road* solo for a couple of years, I am about to violently yank the steering wheel back in a direction that has fewer dependencies. **Those with static build pipeline experience** _now's your time to shine.._  React is optimizing for way too many things that JUST DO NOT MATTER TO OUR USE CASE.  It was an interesting intellectual study, but it ultimately will have a much shorter shelf life than HTDB did (30 years).

----------

Given that, here are the challenges I'd like to have conversations about:

* **Storage of media** - how to make, say, 10G of JBC media (audio/video/pix) be forever accessible alongside the markdown. This  one bugs me the most: only I (currently) have access to and manage the AWS S3 buckets where most of the large media objects are stored. Regardless of where large data ends up (S3, Dropbox, IPFS, archive.org local storage), allowing distributed contributions to that data is a head-scratcher.  Unlike the textual markdown files which are stored IN the JBC repository, large data is outside that model.  I really don't like that.  We at least need to abstract the access layer so that HOW it is stored is a blackbox to the rendering step.
* **Publishing permissions** - we want, say, James Duval to add new pages and edit existing pages.  Markdown and admin approval of code merges - Good enough?
* **Dynamic content** - i've seen solutions whereby new website comments are pushed into the repo and a pipeline does what's necessary to turn them into markdown.  While elegant in that now *ALL* content is truly static and checked-into the codebase, it does rely upon a pretty brittle pipeline.
* **Rendered website** - the output of any markdown pipeline will be a static build/HTML website.  This rendered version could be used by anyone as a backup/offline safety site.  (I have just this week done this to both sumosonic.com and wilsondub.com - their repos are now static codebases)
* **Search** - given the size of all the underlying JSON, I do not think it would be unreasonable for all "searches" to be driven via client-side filtering of local storage copies of the JSON that drives the site.  This would obviate the need for backend APIs to deliver such material. 

----------

Collaboration is key.  Help navigate a sane route. Don't make me stop this computer.

**Comment by hischeapmoves** [comment]
How many pages do you think jbc.com comprises? I've been using Jekyll for one of my sites, and while it's been working fine, it's not the most performant platform and adding a few pages a week has now bogged it down to the point that it takes about 5 minutes to generate the site.

I've been investigating a platform called Astro (https://astro.build) and also been considering Nuxt.

I know exactly what you mean about hosting the larger assets in perpetuity, and search. I'm not sure there's a good technical answer to those issues, but they're something I'm also considering and I'll let you know if I have any promising ideas.

—Ian

**Reply by xpollen8** [reply]
Pages? Depends.

The gigs alone account for nearly 2000.  If you bring in the old mailing list, that's another 3500

Otherwise, maybe 500 for press articles and a couple hundred for albums, lyrics, conspirators.

Let's be pessimistic and say: 1000 to 7000

Does Jeykll truly have to do a full rebuild onchange, or can they be targeted more intelligently?

**Reply by hischeapmoves** [reply]
There is an incremental rebuild now, I believe. I am a major version behind, and when I looked into upgrading to the new Jekyll or learning something new, I chose the latter.

Depending on how the site is rendered, the time it takes may not be a factor. You're just checking in your updates and the site updates some time later, and it's not like you're trying to push out breaking news. Jekyll is certainly easy enough to work in.

I recently put together a mini-wiki with an [Astro-based theme](https://starlight.astro.build/). It has some built-in search, looks like it's an index compiled at build time: https://wiki.ink19.com/

**Reply by hischeapmoves** [reply]
Here's the source for that jekyll site: https://github.com/ubikorp/lacking.org
